### PR TITLE
add starttls (postgres flavour):

### PR DESCRIPTION
**Pull request by maigl**
### What does this PR do?

This adds starttls support which allows to reverse proxy postgres connections through traefik (via SNI).

#### Context:
We have a number of PostgreSQL databases running inside our kubernetes cluster.
All database instances are accessed through traefik via a single IP address. In order to
access the right PostgreSQL instances from the clients we want to use different subdomains.
One subdomain should be the target address of a specific PostgreSQL instance and we want to
route to it via the SNI (server name indication in TLS header).
This approach seems to be the only possible solution in our setup. We don't have any other information
to identify the target PostgreSQL instance from a client request.

#### Problem:
PostgreSQL performs a StartTLS handshake before sending any TLS header. Since traefik doesn't perform
this handshake no TLS header is sent and hence traefik cannot extract the target server name.

#### Idea:
We propose the following steps:

traefik performs StartTLS handshake with client
client now sends TLS header
traefik extracts SNI
traefik performs StartTLS handshake with server
traefik routes connection from client to target PostgreSQL instance

### More

- [x] Added/updated tests
- [ ] Added/updated documentation

### Additional Notes

Other starttls flavors might be added in future. 

<sub>Michael Kuhnt <michael.kuhnt@daimler.com> Daimler TSS GmbH ([Imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md))
</sub>

### CODE REVIEW

**Comment by maigl** [comment]
@matoubidou @MoritzKeppler @sbueringer could you plz review ?

File: pkg/tcp/starttls.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	b := make([]byte, 1)
+	_, err = io.ReadFull(conn, b)
+	if err != nil {
+		return err
+	}
+
+	if b[0] != PostgresStartTLSReply[0] {
+		// we only support secure connections otherwise we wouldn't get any SNI header
+		// and cannot forward.
+		return fmt.Errorf("unexpected postgres starttls response got %v want 'S' ", b)
```
Comment (MoritzKeppler) [inline pkg/tcp/starttls.go:47]:
```suggestion
		return fmt.Errorf("unexpected postgres starttls response got %v want %s ", b, PostgresStartTLSReply[0])
```

File: pkg/tcp/starttls.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	}
+
+	b := make([]byte, 1)
+	_, err = io.ReadFull(conn, b)
+	if err != nil {
+		return err
+	}
+
+	if b[0] != PostgresStartTLSReply[0] {
+		// we only support secure connections otherwise we wouldn't get any SNI header
```
Comment (MoritzKeppler) [inline pkg/tcp/starttls.go:45]: SNI is just mandatory in our case, right?

**Review by MoritzKeppler — COMMENTED** [review]
just two minor things - more (if found) will follow soon.

**Comment by chrischdi** [comment]
in your description:
`This adds postgres starttls postgres support which allows to reverse proxy postgres connection through traefik (via SNI).`
> adds postgres starttls postgres support?!

File: pkg/tcp/starttls.go (hunk 2 above)
Comment (maigl) [reply pkg/tcp/starttls.go:45]: I believe it's mandatory .. but may be I don't get it ?!

File: pkg/server/server_entrypoint_tcp_test.go [hunk 3]
```diff
@@ -221,3 +221,39 @@ func TestReadTimeoutWithFirstByte(t *testing.T) {
...
+
+	_, err = conn.Write(tcp.PostgresStartTLSMsg)
+	require.NoError(t, err)
+
+	buf := make([]byte, 1)
+	_, err = conn.Read(buf)
+	require.NoError(t, err)
+
+	if buf[0] != tcp.PostgresStartTLSReply[0] {
+		t.Fatalf("Expected Start TLS reply but got %v: ", buf[0])
```
Comment (TSSven) [inline pkg/server/server_entrypoint_tcp_test.go:257]:
```suggestion
		t.Fatalf("Expected StartTLS reply but got %v: ", buf[0])
```

File: pkg/tcp/router.go [hunk 4]
```diff
@@ -36,6 +37,18 @@ func (r *Router) ServeTCP(conn WriteCloser) {
 		return
 	}
 
+	// this peeks into the connection and tries to perform the
+	// starttls handshake (postgres). It works in both cases,
```
Comment (TSSven) [inline pkg/tcp/router.go:41]:
```suggestion
	// StartTLS handshake (postgres). It works in both cases,
```

Comment (TSSven) [reply pkg/tcp/router.go:41]: I'm not quite sure what the word group "StartTLS handshake (postgres)" is supposed to say.
Is the handshake postgres-specific, i.e. "StartTLS handshake for postgres"?[1]
Or could this be used in use cases too? If so, the comment should convey this.

[1]
I also checked [rfc2487](https://tools.ietf.org/html/rfc2487) and from what I understood they do call the first activity after executing the STARTTLS command "TLS negotiation". Maybe using this phrase will in some cases make it easier to distinguish and understand for the reader. --> In the example above: "start TLS negotiation for postgres"?

File: pkg/tcp/starttls.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+// HandlePostgresStartTLSHandshakeAsServer performs a StartTLS Handshake (acting as server)
+// It peeks some into some bytes of conn and
+// tries to find out if the client performs a StartTLS handshake.
+// If the client request does contain a StartTLS handshake signature
+// the handshake is performed. After this step the client will
+// start a TLS session.
+// If no StartTLS signature is found the bytes of the connection
+// are unmodified.
```
Comment (TSSven) [inline pkg/tcp/starttls.go:59]:
```suggestion
// remain unmodified.
```

File: pkg/tcp/starttls_test.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+			log.Info("backend: writing response")
+			_, err = conn.Write([]byte(testresponse))
+			require.NoError(t, err)
+		}
+
+	}
+}
+
+// In this test we test the the proxy's configuration against different backends.
+// The proxy must match the the backend. I.e. if backend expect StartTLS we have
```
Comment (TSSven) [inline pkg/tcp/starttls_test.go:73]:
```suggestion
// The proxy must match the the backend. I.e. if the backend expects StartTLS we have
```

File: pkg/tcp/starttls_test.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+// handshake, because this side is handled by the tcp.Router and not the Proxy.
+func TestProxyPostgresStartTLS(t *testing.T) {
+	tests := []struct {
+		name                 string
+		backendWantsStartTLS bool
+		proxyDoesStartTLS    bool
+		wantErr              bool
+	}{
+		{
+			name:                 "proxy  and backend want starttls",
```
Comment (TSSven) [inline pkg/tcp/starttls_test.go:84]:
```suggestion
			name:                 "proxy and backend want starttls",
```

**Review by TSSven — CHANGES_REQUESTED** [review]
My humble contribution... merely a spell check

File: pkg/tcp/router.go [hunk 8]
```diff
@@ -165,6 +178,17 @@ func (r *Router) HTTPSHandler(handler http.Handler, config *tls.Config) {
 	r.httpsTLSConfig = config
 }
 
+// ConnectionFilter allows to add a filter func to prefilter incoming connections.
+// This can be used for example to peek into the connection and perform a StartTLS handshake.
+// The filter func is registered by a given name. Successive calls with the same name will
+// overwrite existing entries.
+func (r *Router) ConnectionFilter(name string, filterFunc func(WriteCloser) (WriteCloser, error)) {
```
Comment (matoubidou) [inline pkg/tcp/router.go:185]: More precisely we are not filtering connections but rather enabling to peek into a connection or to read some bytes before the connection is routed. I would prefer a word reflecting this behaviour. What about `PreRoutingHook`?

**Comment by sbueringer** [comment]
@maigl all findings adressed? Should I also take a look?

**Review by maigl — COMMENTED** [review]
> @maigl all findings adressed? Should I also take a look?

of course .. but I'm not done fixing the findings .. let me finish this first ..

File: pkg/tcp/starttls_test.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+			if withStartTLS {
+				withErr := false
+				buf := make([]byte, len(PostgresStartTLSMsg))
+				log.Info("backend: reading start tls ")
+				if _, err := conn.Read(buf); err != nil {
+					withErr = true
+				}
+
+				if bytes.Equal(buf, PostgresStartTLSMsg) {
+					time.Sleep(1 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (maigl) [reply pkg/tcp/starttls_test.go:36]: don't know how .. maybe we could team up on this .

**Review by maigl — COMMENTED** [review]
> @maigl all findings adressed? Should I also take a look?

@sbueringer yes I think now would be a good time if you could share you thoughts on this! :)

**Comment by maigl** [comment]
outdated .. new try

**Pull request closed without merging**